### ISSUE TITLE
NdOverlay.hist raises AttributeError

### ISSUE THREAD

**Issue by peterroelants** [holoviz/holoviews#5022]
#### Description of expected behavior and the observed behavior 
Creating a histogram from an `NdOverlay` using `hist()` results in an `AttributeError`.

E.g. the line:
```python
nd_overlay.hist(groupby='cat', dimension=['x','y'])
```
raises:
```python-traceback
AttributeError: 'NoneType' object has no attribute 'hist'
```

I would have expected `NdOverlay.hist()` to create a histogram plot similar to the following figure (Not generated by this code):
<img src="https://user-images.githubusercontent.com/951093/125970995-e750d20e-b531-47cb-8a04-62956803c6ef.png"  width="200">
For example, the resulting histogram plot could be a new `NdOverlay` of histograms of the elements in the original `NdOverlay`.

#### Complete, minimal, self-contained example code that reproduces the issue

Reproduce by running:
```python
import numpy as np
import holoviews as hv

hv.extension('bokeh')


points_dct = {
    'p1': hv.Points(np.random.randn(1000, 2), kdims=['x', 'y'], label='p1'),
    'p2': hv.Points(np.random.randn(1000, 2) * 2 + 1, kdims=['x', 'y'], label='p2')
}
nd_overlay = hv.NdOverlay(points_dct, kdims='cat')
nd_overlay.hist(groupby='cat', dimension=['x','y'])
```

#### Stack traceback and/or browser JavaScript console output
The line `nd_overlay.hist(groupby='cat', dimension=['x','y'])` results in the following error output:
```python-traceback
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/tmp/ipykernel_58893/2897274094.py in <module>
      8 }
      9 nd_overlay = hv.NdOverlay(points_dct, kdims='cat')
---> 10 nd_overlay.hist(groupby='cat', dimension=['x','y'])

~/miniconda3/envs/lcms/lib/python3.9/site-packages/holoviews/core/overlay.py in hist(self, dimension, num_bins, bin_range, adjoin, index, **kwargs)
     75             raise TypeError("Please supply a suitable index or label for the histogram data")
     76 
---> 77         hists = self.get(index).hist(
     78             adjoin=False, dimension=dimension, bin_range=bin_range,
     79             num_bins=num_bins, **kwargs)

AttributeError: 'NoneType' object has no attribute 'hist'
```

It seems that the default index is `0`, which is not in the NdOverlay. From the debugger:
```python-traceback
> /home/peter/miniconda3/envs/lcms/lib/python3.9/site-packages/holoviews/core/overlay.py(77)hist()
     75             raise TypeError("Please supply a suitable index or label for the histogram data")
     76 
---> 77         hists = self.get(index).hist(
     78             adjoin=False, dimension=dimension, bin_range=bin_range,
     79             num_bins=num_bins, **kwargs)

ipdb>  index
0
ipdb>  [el for el in self]
[:Points   [x,y], :Points   [x,y]]
ipdb>  self.keys()
['p1', 'p2']
ipdb>  repr(self)
':NdOverlay   [cat]\n   :Points   [x,y]'
```

#### ALL software version info
```
Python version       : 3.9.6

bokeh    : 2.3.3
holoviews: 1.14.4
numpy    : 1.21.0

jupyter core     : 4.7.1
jupyter-notebook : 6.4.0
qtconsole        : not installed
ipython          : 7.25.0
ipykernel        : 6.0.1
jupyter client   : 6.1.12
jupyter lab      : 3.0.16
nbconvert        : 6.1.0
ipywidgets       : 7.6.3
nbformat         : 5.1.3
traitlets        : 5.0.5
```

**Comment by peterroelants** [comment]
For anyone having the same issue: As a workaround, it is possible to create the histograms manually, e.g.:

```
hist_dct = {
    k: hv.operation.histogram(p, dimension='x')
    for k, p in points_dct.items()
}
hist_nd_overlay = hv.NdOverlay(hist_dct, kdims=['cat'])
```

**Comment by jbednar** [comment]
What were you expecting that calling .hist() would do on an overlay? Normally it does an adjoined histogram by default, so this is what I'd hope would happen:

![image](https://user-images.githubusercontent.com/1695496/125966944-33df7969-ad37-481a-a1d5-ee8b24e13303.png)

**Comment by peterroelants** [comment]
Hi, @jbednar I would have expected the `nd_overlay.hist(groupby='cat', dimension=['x','y'])` line to create a histogram plot of the elements in `nd_overlay`:

<img src="https://user-images.githubusercontent.com/951093/125970995-e750d20e-b531-47cb-8a04-62956803c6ef.png"  width="250">

I wouldn't have expected to be an Adjoint, however, I would have expected it to be usable in an AdjointLayout using the `<<` operator as you illustrated.

Just to be clear, the following line does not work and gives the `AttributeError` error:
```python
nd_overlay.hist(groupby='cat', dimension=['x','y'])
```

**Comment by jbednar** [comment]
Ok, if you wanted that, I think your original call should have been `nd_overlay.hist(groupby='cat', adjoin=False, dimension=['x','y'])`, and I think you should have gotten the output I showed with your actual call `nd_overlay.hist(groupby='cat', dimension=['x','y'])` (because `adjoin=True` by default).  Of course, you got neither one, just an error, and I'd love to see a patch that gives an overlaid adjoint or non-adjoint histogram plot (as appropriate) when given an overlay; seems very useful!

**Comment by peterroelants** [comment]
It seem that the regular `hv.Overlay.hist()` also has some unexpected behaviour. It only returns the histogram of the first element. I think these issues might be related to the overlays not being mapped over to create a new overlay of histogram elements.

E.g. the following:
```python
import numpy as np
import holoviews as hv

hv.extension('bokeh')


pts1 = hv.Points(np.random.randn(1000, 2) , kdims=['x', 'y'], label='p1l')
pts2 = hv.Points(np.random.randn(1000, 2) - 5, kdims=['x', 'y'], label='p2l')
pts_overlay = hv.Overlay([pts1, pts2])   # pts_overlay = pts1 * pts2
pts_overlay.hist(dimension=['x','y'])
```
Results in:
<img src="https://user-images.githubusercontent.com/951093/126057785-78dc68f9-2661-489d-8097-44d4cf3fb8a5.png"  width="300">

**Comment by peterroelants** [comment]
Hi @jbednar , I made a PR that fixes the original `AttributeError` issue with the `NdOverlay`, and makes the `CompositeOverlay.hist()` function more general by showing overlaid histograms of all elements in the original overlay.

I would love to get your thoughts on the PR at https://github.com/holoviz/holoviews/pull/5031

**Comment by peterroelants** [comment]
Fixed by https://github.com/holoviz/holoviews/pull/5031 (which is on lasted master and will only be released from v2.0)
